test/unit/parallel_pusher_test.rb

```diff
@@ -0,0 +1,43 @@
+require 'test_helper'
+
+class ParallelPusherTest < ActiveSupport::TestCase
+  self.use_transactional_tests = false
+
+  context "when pushing gems in parallel" do
+    setup do
+      @fs = RubygemFs.mock!
+      @user = create(:user, email: "user@example.com")
+    end
+
+    teardown do
+      @user.destroy
+      @rubygem = Rubygem.find_by(name: 'hola')
+      @rubygem.versions.destroy_all
+      @rubygem.destroy
+      Delayed::Job.delete_all
+      GemDownload.delete_all
+    end
+
+    should "not lead to sha mismatch between gem file and db" do
+      @gem1 = gem_file("hola-0.0.0.gem")
+      @gem2 = gem_file("hola/hola-0.0.0.gem")
+      @cutter1 = Pusher.new(@user, @gem1)
+      @cutter2 = Pusher.new(@user, @gem2)
+
+      t1 = Thread.new do
```

Comment:
Isn't this threading introducing random order of execution of gem processing making this test "unstable"? Isn't there any chance to use `ActiveSupport Latch` to control the order?

Reply:
yes, it would definitely make it better. Thanks @simi 
`ActiveSupport::Concurrency::Latch` was deprecated in 5.1 and was replaced by `[Concurrent::CountDownLatch](http://ruby-concurrency.github.io/concurrent-ruby/Concurrent/CountDownLatch.html)` https://github.com/rails/rails/pull/20866. I have updated test to use the same.

Reply:
I don't see any difference with current code and the old one (using thread join). I think you should be able to use `Latch` to stop thread 1 somewhere, start thread 2 and continue in thread 1 to simulate the exact order of critical operations resulting in race condition at some points (which is the reason of this PR). Maybe I'm wrong. I try to submit my code to compare if welcomed.

Reply:
>  I try to submit my code to compare if welcomed.

Yes, please :pray: Even an example code of what you are trying to say would be fine as well.
I didn't even know about `ActiveSupport::Concurrency::Latch` until you mentioned it.